cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/ReadSegmentRangeCommand.java

```diff
@@ -42,29 +47,97 @@ public ReadSegmentRangeCommand(CommandArgs args) {
+    }
+
+    /**
+     * Creates the file (and parent directory if required) into which thee segment data is written.
+     *
+     * @param fileName
+     * @return A {@link File} object representing the filename provided.
+     * @throws FileAlreadyExistsException if the file already exists, to avoid any accidental overwrites.
+     * @throws IOException if the file/directory creation fails.
+     */
+    private File createFileAndDirectory(String fileName) throws IOException {
+        File f = new File(fileName);
+        // If file exists throw FileAlreadyExistsException, an existing file should not be overwritten with new data.
+        if (f.exists()) {
+            throw new FileAlreadyExistsException("Cannot write segment data into a file that already exists.");
+        }
+        if (!f.getParentFile().exists()) {
+            f.getParentFile().mkdirs();
+        }
+        f.createNewFile();
+        return f;
+    }
+
+    /**
+     * Reads the contents of the segment starting from the given offset and writes into the provided file.
+     *
+     * @param segmentHelper             A {@link SegmentHelper} instance to read the segment.
+     * @param segmentStoreHost          Address of the segment-store to read from.
+     * @param fullyQualifiedSegmentName The name of the segment.
+     * @param offset                    The starting point from where the segment is to be read.
+     * @param length                    The number of bytes to read.
+     * @param fileName                  A name of the file to which the data will be written.
+     * @throws IOException if the file create/write fails.
+     * @throws InterruptedException if the request fails.
+     * @throws ExecutionException if the request fails.
+     * @throws TimeoutException if the request fails.
+     */
+    private void readAndWriteSegmentToFile(SegmentHelper segmentHelper, String segmentStoreHost, String fullyQualifiedSegmentName,
+                                           int offset, int length, String fileName) throws IOException, InterruptedException, ExecutionException, TimeoutException {
+        File file = createFileAndDirectory(fileName);
+
+        output("Downloading %d bytes from offset %d into %s.", length, offset, fileName);
+        int currentOffset = offset;
+        int bytesToRead = length;
+        int progress = 0;
+
+        while (bytesToRead > 0) {
+            int bufferLength = Math.min(READ_WRITE_BUFFER_SIZE, bytesToRead);
+            CompletableFuture<WireCommands.SegmentRead> reply = segmentHelper.readSegment(fullyQualifiedSegmentName,
+                    currentOffset, bufferLength, new PravegaNodeUri(segmentStoreHost, getServiceConfig().getAdminGatewayPort()), "");
+            WireCommands.SegmentRead bufferRead = reply.get(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            // Write the buffer into the file.
+            try (FileOutputStream fileOutputStream = new FileOutputStream(file, true)) {
```

Comment:
You don't need to open the `FileOutputStream` in every iteration; open it outside of the `while` loop; you can always reuse it - and make your code more efficient at the same time.

Reply:
Opened the `FileOutputStream` outside the loop with an `@Cleanup` annotation.